core/src/services/ocios/core.rs

```diff
@@ -0,0 +1,309 @@
+            .field("namespace", &self.namespace)
+            .finish_non_exhaustive()
+    }
+}
+
+impl OciOsCore {
+    async fn load_credential(&self) -> Result<Option<OCICredential>> {
+        let cred = self
+            .loader
+            .load()
+            .await
+            .map_err(new_request_credential_error)?;
+
+        if let Some(cred) = cred {
+            Ok(Some(cred))
+        } else {
+            // Mark this error as temporary since it could be caused by Aliyun STS.
+            Err(Error::new(
+                ErrorKind::PermissionDenied,
+                "no valid credential found, please check configuration or try again",
+            )
+            .set_temporary())
+        }
+    }
+
+    pub async fn sign<T>(&self, req: &mut Request<T>) -> Result<()> {
+        let cred = if let Some(cred) = self.load_credential().await? {
+            cred
+        } else {
+            return Ok(());
+        };
+
+        self.signer.sign(req, &cred).map_err(new_request_sign_error)
+    }
+
+    #[inline]
+    pub async fn send(&self, req: Request<AsyncBody>) -> Result<Response<IncomingAsyncBody>> {
+        self.client.send(req).await
+    }
+}
+
+impl OciOsCore {
+    pub fn os_get_object_request(
+        &self,
+        path: &str,
+        _range: BytesRange,
+        _is_presign: bool,
+        _if_match: Option<&str>,
+        _if_none_match: Option<&str>,
+        _override_content_disposition: Option<&str>,
+    ) -> Result<Request<AsyncBody>> {
+        let p = build_abs_path(&self.root, path);
+        let url = format!(
+            "objectstorage.{}.oraclecloud.com/n/{}/b/{}/o/{}",
+            self.region,
+            self.namespace,
+            self.bucket,
+            percent_encode_path(&p)
+        );
+
+        let req = Request::get(&url)
+            .body(AsyncBody::Empty)
+            .map_err(new_request_build_error)?;
+
+        Ok(req)
+    }
+
+    pub fn os_head_object_request(
+        &self,
+        path: &str,
+        _is_presign: bool,
+        _if_match: Option<&str>,
+        _if_none_match: Option<&str>,
+    ) -> Result<Request<AsyncBody>> {
+        let p = build_abs_path(&self.root, path);
+        let url = format!(
+            "objectstorage.{}.oraclecloud.com/n/{}/b/{}/o/{}",
+            self.region,
+            self.namespace,
+            self.bucket,
+            percent_encode_path(&p)
+        );
+
+        let req = Request::head(&url)
+            .body(AsyncBody::Empty)
+            .map_err(new_request_build_error)?;
+
+        Ok(req)
+    }
+
+    pub fn os_list_object_request(
+        &self,
+        path: &str,
+        _delimiter: &str,
+        _limit: Option<usize>,
+        _start_after: Option<String>,
+    ) -> Result<Request<AsyncBody>> {
+        let p = build_abs_path(&self.root, path);
+        let url = format!(
+            "objectstorage.{}.oraclecloud.com/n/{}/b/{}/o",
+            self.region, self.namespace, self.bucket,
```

Comment:
If the endpoint is always the same for all API, we can build this endpoint during build.